Add 'Take Sourcegraph to your [code host]' popover

tjkandala opened the pull request:
Closes #14199. Part of [RFC 221](https://docs.google.com/document/d/19f4xleYBU1zZZdqMmXlLmFxeR-fwEpOwTOgViOFOnyo/)

- [x] - Convert `GoToCodeHostAction` to function component
- [x] - Implement ~popup~ popover
- [x] - Remove chrome extension toast 
- [x] - Implement focus trapping for modal
- [x] - Implement hover tracking (5 hovers seen by user before displaying popup)
- [x] - Storybook
- [x] - Integration testing
- [x] - Improve CSS
    - [x] - SVG sizing
    - [x] - Popover arrow

#### Interaction
<details>
<summary>Popover fade in on click</summary>
<img src="https://user-images.githubusercontent.com/37420160/94746927-d374a680-034b-11eb-848d-ae54a0ce9200.gif" />
</details>

<details>
<summary>Keyboard navigation, "Remind me later"</summary>
<img src="https://user-images.githubusercontent.com/37420160/94747682-8c87b080-034d-11eb-9942-a05f62f322c4.gif" />
</details>

#### Popover visuals

![Screenshot from 2020-09-30 19-02-49](https://user-images.githubusercontent.com/37420160/94748554-93afbe00-034f-11eb-82cd-414adf72db3b.png)

@AlicjaSuska The "arrow" is a square rotated 45 degrees and 50% hidden by the popover. This approach is good for responsive popovers since it's mostly the same styles for each direction, but it doesn't align 100% with the design. I'll revisit the arrow after finishing the remaining issues.

sqs commented:
In https://github.com/sourcegraph/sourcegraph/pull/14261 I made it so it shows the toast on your 3rd day of activity, not your 1st. Not sure how you're doing it here, but you should make it so this doesn't show up the very first time they visit Sourcegraph.

tjkandala commented:
We're removing the toast as part of [RFC 221](https://docs.google.com/document/d/19f4xleYBU1zZZdqMmXlLmFxeR-fwEpOwTOgViOFOnyo) in favor of inline alerts after the user has seen 2 hovers.

That's a good idea though; cc @sourcegraph/web @AlicjaSuska Should we check `daysActiveCount` for this popup?

AlicjaSuska commented:
This popup is displayed after clicking on 'view on GitHub'. I think it doesn't make sense to add the day count as criteria here. I think it makes sense to show this information even if someone clicks 'view on [code host]' for the first time.

felixfbecker commented:
@AlicjaSuska maybe we can include the criterium "Has seen at least 2 hovers" too with the same reasoning as for the alert?

felixfbecker commented on web/src/auth/icons.tsx:
```diff
@@ -1,7 +1,17 @@
+    ...props
+}) => (
+    <svg
+        width={size ?? '65'}
+        height={size ?? '64'}
```
**suggestion**: a `size` prop doesn't really make sense to me given this SVG doesn't even seem to be square. Is there any reason we can't use CSS to set the `height` or `width`? That way we can use `rem`, enforce our stylelint rules, easily make it responsive if needed etc. I'd generally avoid `size` because it's essentially an inline style like `style={{ height: '65px' }}`, which we disallow in favor of using CSS and class names.

felixfbecker replied:
This change wasn't reverted

felixfbecker commented on web/src/repo/actions/GoToCodeHostAction.tsx:
```diff
@@ -26,133 +31,205 @@ interface Props extends RevisionSpec, Partial<FileSpec> {
-        )
-    }
+            {modalOpen && (
+                <ModalContainer onClose={onClose} hideCloseIcon={true} className="justify-content-center">
```
**suggestion**: I'd put the modal into its own component (I was a bit confused for example by class names like `code-host-action__plus-icon`)

felixfbecker commented on web/src/repo/actions/GoToCodeHostAction.tsx:
```diff
@@ -26,133 +31,205 @@ interface Props extends RevisionSpec, Partial<FileSpec> {
+                            </div>
+
+                            <div className="d-flex justify-content-end">
+                                <LinkOrButton
```
**question**: this is always a link, so why use `LinkOrButton` instead of `<a>`? Reading `LinkOrButton`, it would imply to the reader that it's dynamic, which could be confusing.

felixfbecker replied:
Okay, I just figured out that `LinkOrButton` is really badly named. It's not a link _or_ a button in any way (like `LinkOrSpan`), it's _always_ a link, but with the _behavior_ of a button. Imo it should be renamed to `ButtonLink`.

felixfbecker commented on web/src/repo/actions/GoToCodeHostAction.tsx:
```diff
@@ -26,133 +31,205 @@ interface Props extends RevisionSpec, Partial<FileSpec> {
+                            </p>
+
+                            <div className="mx-auto code-host-action__graphic-container d-flex justify-content-between align-items-center">
+                                <SourcegraphIcon size={48} />
+                                <PlusThickIcon size={20} className="code-host-action__plus-icon" />
+                                <Icon size={56} />
```
**suggestion**: instead of unitless sizes, could we define this in CSS?

felixfbecker reviewed (commented):
**issue**: How are we going to test this and prevent regressions? Having a story for the modal sounds like a good idea, and maybe an integration test for the behavior?

AlicjaSuska commented:
@felixfbecker let's add the '2 hovers' criteria, you're right.

tjkandala commented:
If the user clicks "Install the Sourcegraph browser extension" from the alert, but doesn't install the extension, should we disable the popover for the 'View on code host button'? They're both dependent on the user seeing 5 hovers now, so it might be annoying to see the popover after already looking to, but not installing, the extension.

tjkandala commented on web/src/repo/actions/GoToCodeHostAction.tsx:
```diff
@@ -3,156 +3,294 @@ import { upperFirst } from 'lodash'
+            targetID={targetID}
+            popperOptions={useMemo(
+                () => ({
+                    placement: 'bottom-start' as const,
```
const assertion isn't necessary here

tjkandala commented on web/src/integration/blob-viewer.test.ts:
```diff
@@ -215,5 +215,45 @@ describe('Blob viewer', () => {
+
+                for (let index = 0; index < HOVER_THRESHOLD; index++) {
+                    await driver.page.click(index % 2 === 0 ? '.test-log-token' : '.test-console-token')
+                    await driver.page.waitForSelector('.hover-overlay', { visible: true })
```
Should I keep this check? The test passes without it, but we may need to wait for the hover overlay if the hover provider in our mock extension becomes asynchronous.

felixfbecker replied:
Everything in Puppeteer is always asynchronous, so yes, we should always have such checks.

tjkandala commented on web/src/integration/blob-viewer.test.ts:
```diff
@@ -215,5 +215,45 @@ describe('Blob viewer', () => {
+                const popoverHeader = await driver.page.evaluate(
+                    () => document.querySelector('.test-install-extension-popover-header')?.textContent
+                )
+                assert.deepStrictEqual(popoverHeader, "Take Sourcegraph's code intelligence to GitHub!")
```
Should we also assert that the popover is _not_ visible before $HOVER_OVERLAY hovers? If so, what would be the best approach here (I wasn't able to find such an assertion in our integration tests).

tjkandala replied:
Is asserting that `execute(() => document.querySelector('.test-install-extension-popover')` returns null a satisfactory solution?

felixfbecker replied:
I think we should have such an assertions and I think that is a perfectly fine way to verify that.

AlicjaSuska commented:
@tjkandala 
- I've changed the pointer to 12px square rotated 45 degrees - as you've suggested [See the design](https://www.figma.com/file/V5t7KOzC6PIX75L4dz8IFc/RFC-221-Browser-extension-improve-discoverability-and-awareness-Approved?node-id=378%3A28)
- I've updated the designs in [Figma](https://www.figma.com/file/V5t7KOzC6PIX75L4dz8IFc/RFC-221-Browser-extension-improve-discoverability-and-awareness-Approved?node-id=378%3A28). Please, pay attention to [dark mode](https://www.figma.com/file/V5t7KOzC6PIX75L4dz8IFc/RFC-221-Browser-extension-improve-discoverability-and-awareness-Approved?node-id=507%3A273) - I've changed the background color slightly
- about your comment: _'If the user clicks "Install the Sourcegraph browser extension" from the alert, but doesn't install the extension, should we disable the popover for the 'View on code host button'?_ - I see the value in showing the popover in the context of clicking on 'view on [code host] even if user already visited the extension page but didn't install. It can serve as reminder / seeing the potential benefit in context. I think they shouldn't be dependent. Both should be dismissed only explicitly by the user (by X for alert or 'no, thanks' for the popover) or by installing the extension.

felixfbecker commented on shared/src/components/LinkOrButton.tsx:
```diff
@@ -34,16 +34,17 @@ interface Props extends Pick<AnchorHTMLAttributes<never>, 'target' | 'rel'> {
  *
- * It is keyboard accessible: unlike <Link> or <a>, pressing the enter key triggers it. Unlike
- * <button>, it shows a focus ring.
+ * It is keyboard accessible: unlike `<Link>` or `<a>`, pressing the enter key triggers it.
```
**thought**: I'm starting to question the validity of the entire `LinkOrButton`/`ButtonLink`  component. This comment stated that `<button>`s don't show a focus ring, which is false - the buttons in our app all have focus rings. But now I just verified that hitting <kbd>enter</kbd> when a link is focused also triggers navigation in Chrome, so that statement seems to be false too, leaving no reason to use this component.
The only case where it would not do anything is if the link only had `onClick`, but in that case it _should not be_ a link, it should be a `<button>`. It seems like for our use cases we should always use `<button>` or `<a>`/`<Link>` directly and avoid this component, potentially remove it.
cc @sourcegraph/frontend-engineers

felixfbecker commented on web/src/repo/actions/InstallExtensionPopover.tsx:
```diff
@@ -0,0 +1,89 @@
+                <div
+                    ref={modalBodyReference as React.MutableRefObject<HTMLDivElement>}
+                    // TODO: move .extension-permission-modal to .modal-body
+                    className="extension-permission-modal p-4 web-content text-wrap border shadow test-install-extension-popover"
```
**suggestion**: Bootstrap has CSS popovers, including triangles. Should we just use those (CSS only)? https://getbootstrap.com/docs/4.5/components/popovers/
If they don't match the design exactly, there are likely Sass variables we can set to refine colors etc.

tjkandala replied:
I'm not sure that we can use it [CSS-only](https://getbootstrap.com/docs/4.5/components/popovers/#example-enable-popovers-everywhere), and the [API seems more complex](https://getbootstrap.com/docs/4.5/components/popovers/#methods) than raw Popper. I'll try using Reactstrap's `Popover`, which looks [promising](https://reactstrap.github.io/components/popovers/) ([implementation](https://github.com/reactstrap/reactstrap/blob/master/src/TooltipPopoverWrapper.js)).

felixfbecker replied:
Curious what happens when an element is assigned the CSS class `popover`? Why doesn't it work?

tjkandala replied:
Nothing happens :). I think we have to activate them with this [jQuery incantation](https://getbootstrap.com/docs/4.5/components/popovers/#example-enable-popovers-everywhere). We haven't had to do this yet since we use reactstrap's tooltip component, which re-implements this functionality without jQuery.

It looks like Bootstrap [adds the tooltip method to jQuery](https://sourcegraph.com/github.com/twbs/bootstrap/-/blob/dist/js/bootstrap.bundle.js#L6681), so we wouldn't be able to access it.

felixfbecker replied:
I wouldn't expect any of the behavior to work, but I would expect it to apply some styling (even if e.g. it's hidden by default). It seems that it doesn't apply anything, which led me to check whether we actually import the popover styles from Bootstrap yet in SourcegraphWebApp.scss, and we don't. So that explains why nothing happens. If we import it, there is styling applied, although the arrow is not positioned correctly without JS (e.g. using Reactstrap or custom logic). The CSS could potentially save the custom CSS that's in this PR for popovers and the TODO for making arrows.

felixfbecker replied:
What I was trying to understand was the reason behind why you said "I'm not sure that we can use it CSS-only"

tjkandala commented on web/src/components/useModality.ts:
```diff
@@ -0,0 +1,124 @@
+
+/**
+ * Utility hook for modal-type components (e.g. ModalContainer, PopoverContainer).
+ * `useModality` adds focus trapping and intuitive close logic to modal containers.
```
It looks like it may be removed from the spec based on [this comment](https://github.com/whatwg/html/issues/3567#issuecomment-451451406). Regardless, the polyfill looks good, so I'll look into it after implementing the rest of RFC 221.

tjkandala commented on web/src/components/PopoverContainer.scss:
```diff
@@ -0,0 +1,53 @@
+.popover-container {
+    z-index: $zindex-popover;
+
+    animation: fadeIn 0.3s both ease-in;
```
I don't think that would work here ([explanation](https://fvsch.com/transition-fade)). 

This works for `HoverOverlay` since we set the opacity to 0 before we get the `overlayPosition`.
![Screenshot from 2020-10-02 17-05-15](https://user-images.githubusercontent.com/37420160/94969873-7a367f80-04d1-11eb-900a-c62119cba55b.png)

tjkandala commented:
Update: Replaced all (2) modals in the web app with `@reach/dialog`, replaced custom popover with reactstrap's popover and `@reach/popover` (to compare), and removed `useModality` hook.

### Reactstrap popover

<details>
<summary>Keyboard navigation</summary>
<img src="https://user-images.githubusercontent.com/37420160/95006224-05963a80-05d0-11eb-9276-90ec85b1e2f2.gif" />
</details>

Fade in and visuals are same as in original post.

### Reach popover

<details>
<summary>Keyboard navigation</summary>
<img src="https://user-images.githubusercontent.com/37420160/95006425-5e66d280-05d2-11eb-9559-867fff53b4ce.gif" />
</details>


No fade in.

#### Popover visuals

![Screenshot from 2020-10-03 23-46-01](https://user-images.githubusercontent.com/37420160/95006450-a259d780-05d2-11eb-911e-1032305d0183.png)

#### Summary

Reactstrap's popover is easier to visually customize, but I don't think it's meant to be used with interactive elements. Reach adheres to the [ARIA spec](https://www.w3.org/TR/wai-aria-practices-1.2/#dialog_modal), but is hard to position when compared to Popper's `modifiers` system (Popper does the heavy lifting for both reactstrap's popover and the original custom popover). 

I'll spend a little more time trying to get Reach's popover to align with the design. If that doesn't work out, I propose that we use reactstrap's popper combined with the [focus-trap](https://github.com/focus-trap/focus-trap) library to essentially recreate the original implementation, but with more robust and mature 3rd-party building blocks.

lguychard commented on web/src/repo/actions/GoToCodeHostAction.tsx:
```diff
@@ -3,159 +3,216 @@ import { upperFirst } from 'lodash'
+     * The external links for the current file/dir, or undefined while loading, null while not
+     * needed (because not viewing a file/dir), or an error.
+     */
+    const [nextComponentUpdate, fileExternalLinksOrError] = useEventObservable<
```
Isn't getting an observable based off props changes meant to be achieved with dependency arrays?

```typescript
    /**
     * The external links for the current file/dir, or undefined while loading, null while not
     * needed (because not viewing a file/dir), or an error.
     */
    const fileExternalLinksOrError = useObservable<ExternalLinkFields[] | null | undefined | ErrorLike>(
        useMemo(() => {
            if (!props.repo || !props.filePath) {
                return of(null)
            }
            return merge(
                of(undefined),
                fetchFileExternalLinks({ repoName: props.repo.name, revision, filePath: props.filePath }).pipe(
                    catchError(error => [asError(error)])
                )
            )
        }, [props.repo, props.filePath])
    )
```

felixfbecker commented on web/src/repo/actions/InstallExtensionPopover.story.tsx:
```diff
@@ -15,17 +15,32 @@ const { add } = storiesOf('web/repo/actions', module).addDecorator(story => (
-            />
-        )}
+        {() => {
+            const [open, setOpen] = useState(false)
+            useEffect(() => {
+                setTimeout(() => setOpen(true), 0)
+            }, [])
```
It would be good to add a comment why this is needed (it's not obvious)

felixfbecker commented:
@tjkandala looks like you did good research on the best solution for popovers! If we're using Reactstrap now, does that mean we also need to add an import to the Bootstrap styles and they should "just work"? (I see that we still have custom CSS for the popover despite using Reactstrap, and that the arrow still doesn't work)

tjkandala commented:
> @tjkandala looks like you did good research on the best solution for popovers! If we're using Reactstrap now, does that mean we also need to add an import to the Bootstrap styles and they should "just work"? (I see that we still have custom CSS for the popover despite using Reactstrap, and that the arrow still doesn't work)

Replaced custom CSS with Bootstrap styles, which fixed the arrow. Weirdly, importing Bootstrap styles changed the way that the offset works. 

#### No offset
![Screenshot from 2020-10-05 18-56-35](https://user-images.githubusercontent.com/37420160/95144700-482e5300-0747-11eb-8559-778b09dae3e4.png)

#### Offset "0, -4" (Same as original post)
![Screenshot from 2020-10-05 18-43-11](https://user-images.githubusercontent.com/37420160/95144717-51b7bb00-0747-11eb-99c2-5aafc3785110.png)

AlicjaSuska reviewed (changes requested):
Thank you for this work @tjkandala I finally could test it after your help 🎉

Improvements that are needed:
- popup should appear for users who have **3 hovers**, not 5 (as it is now)
- **pointer** styles are not aligned with the design. If it's a very difficult fix we can also remove the pointer entirely to speed things up.
- in dark mode, there are **light points** in the corners of the popover. I'm not sure what are those.
<img width="439" alt="Screenshot 2020-10-06 at 18 40 36" src="https://user-images.githubusercontent.com/20326070/95231895-10d1ac00-0804-11eb-8470-c8ae9103a4f9.png">
- if we move forward with the pointer, it should be closer to the icon, 4 px above the divider (now it touches the divider) Design screenshot:
<img width="147" alt="Screenshot 2020-10-06 at 18 46 07" src="https://user-images.githubusercontent.com/20326070/95232076-50989380-0804-11eb-945f-b7628f7fbd07.png">
- after clicking on 'install the extension', Docs should open in the **new tab** (not in the same tab as currently)

tjkandala commented:
![Screenshot from 2020-10-06 15-20-47](https://user-images.githubusercontent.com/37420160/95253721-cf81d200-07ec-11eb-924f-d9c953acf704.png)

update: @AlicjaSuska. Links also open in new tabs when the user originally middle clicked 'Go to [code host]'.

felixfbecker commented on web/src/components/Dialog.scss:
```diff
@@ -0,0 +1,17 @@
+@import '~@reach/dialog/styles.css';
+
+.modal-body {
```
**thought**: Bootstrap has `modal` too, I wonder if we should reuse that (not in this PR or iteration though)

tjkandala replied:
+1, Bootstrap modals are good

felixfbecker commented on web/src/integration/blob-viewer.test.ts:
```diff
@@ -215,5 +216,53 @@ describe('Blob viewer', () => {
+                const popoverHeader = await driver.page.evaluate(
+                    () => document.querySelector('.test-install-extension-popover-header')?.textContent
+                )
+                assert.deepStrictEqual(
```
```suggestion
                assert.strictEqual(
```

felixfbecker commented on web/src/repo/RepoContainer.tsx:
```diff
@@ -305,6 +317,33 @@ export const RepoContainer: React.FunctionComponent<RepoContainerProps> = props
+        return false
+    })
+
+    /** TODO description */
```
TODO

felixfbecker commented on web/src/repo/actions/InstallExtensionPopover.scss:
```diff
@@ -0,0 +1,29 @@
+    }
+
+    & .arrow::after {
+        border-bottom-color: var(--body-bg) !important;
```
**question**: shouldn't this be set in global-styles for all popovers? Or perhaps there is just a Bootstrap Sass var we can set for this? Usually we need to set some Sass variables for every Bootstrap component when start using it to make it theme-aware, and this PR is the first one to use `popover`, so that's likely

tjkandala replied:
Implemented this change. However, I found out that other popovers in our application broke once we imported Bootstrap's popover styles, probably since they were built without Bootstrap in mind. I set `$popover-max-width` to `auto`. Do you think this is the right change?

felixfbecker replied:
We obviously can't break other popovers, so if that change fixes it, that sounds right. I wasn't aware we actually used the class elsewhere given we had not imported it. I also just found these styles:

https://github.com/sourcegraph/sourcegraph/blob/9308b5e932036efde80e6a7b926383130e8aeb98/web/src/SourcegraphWebApp.scss#L233-L239

They can probably be removed, and that comment looks like it doesn't apply if we set the [variables](https://sourcegraph.com/search?q=repo:%5Egithub%5C.com/twbs/bootstrap%24+%24popover+-f:dist&patternType=literal) correctly. The arrow also seems to be an element you have to explicitly render in the DOM, so that can just be omitted if we don't want it for certain popovers (like the rev switcher).

felixfbecker commented on web/src/repo/actions/InstallExtensionPopover.tsx:
```diff
@@ -27,11 +28,13 @@ export const InstallExtensionPopover: React.FunctionComponent<CodeHostExtensionP
     const { displayName, icon } = serviceTypeDisplayNameAndIcon(serviceType)
     const Icon = icon || ExportIcon
 
-    // TODO: store type of original click (normal vs aux). If it was an aux click, open link in new tab
```
**suggestion**: shouldn't we just always open in a new tab, since it's an external link? Seems like that would save a lot of code

tjkandala replied:
I implemented this since the original implementation opened the link in the same tab, but you're right. I'll revert this

felixfbecker commented:
> popup should appear for users who have 3 hovers, not 5 (as it is now)

@AlicjaSuska this was something we discussed when we hacked on this, 3 felt very low. Do you feel strongly about 3?

tjkandala commented on package.json:
```diff
@@ -259,6 +259,7 @@
     "@hot-loader/react-dom": "^16.13.0",
     "@primer/octicons-react": "^9.6.0",
     "@reach/accordion": "^0.10.2",
+    "@reach/dialog": "^0.11.2",
```
We use `@reach/dialog` for `SavedSearchModal` and `ExtensionPermissionModal`

AlicjaSuska commented:
> > popup should appear for users who have 3 hovers, not 5 (as it is now)
> 
> @AlicjaSuska this was something we discussed when we hacked on this, 3 felt very low. Do you feel strongly about 3?

I think 3 is enough to understand what tooltips do. It's a gut feeling though, I don't have any strong evidence of 3 being better than 5. Why 3 felt low for you? @felixfbecker

felixfbecker commented:
Because you could reach 3 very easily accidentally by just moving the mouse across the page

felixfbecker commented:
@AlicjaSuska I think your concerns were all adressed, if not we can follow up :)

@tjkandala let's merge 🎉 
You can rebase with `git rebase --strategy-option="rename-threshold=10" main` which handled the renames flawlessly for me.
